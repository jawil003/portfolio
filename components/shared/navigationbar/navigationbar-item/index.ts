export { NavigationBarItem } from "./navigationbar-item";
export type { Props as NavigationBarItemProps } from "./navigationbar-item";
