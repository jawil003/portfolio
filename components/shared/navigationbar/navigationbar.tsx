import {
  motion,
  useAnimation,
} from "framer-motion";
import { memo, useState } from "react";
import { css } from "@emotion/react";
import { useNavigationIcons } from "../../../hooks/useIcons.hook";
import { Logo } from "../logo/logo";
import { NavigationBarItem } from "./navigationbar-item";
import { FlexContainer } from "../flexcontainer/flexcontainer";
import Spacer from "../spacer/spacer";
import {
  header as items,
  footer as footerItems,
} from "../../../config/routes.json";
import HamburgerMenuIcon from "../../../icons/hamburgerMenu.icon";
import designSystem from "../../../styles/designSystem";
import DesktopNavigationBarContext from "../../../contexts/DesktopNavigationBarContext";

export interface Props {
  className?: string;
  backgroundIsWhite?: boolean;
}

/**
 * An NavigationBar React Component.
 * @author Jannik Will
 * @version 0.1
 */
export const NavigationBar: React.FC<Props> = memo(
  ({
    className,
    backgroundIsWhite,
  }) => {
    const getIcon = useNavigationIcons();
    const navBar = useAnimation();
    const headerBar = useAnimation();
    const [open, setOpen] = useState(
      false,
    );

    return (
      <>
        <DesktopNavigationBarContext.Provider
          value={{
            open,
            setOpen: (
              openState: boolean,
            ) => setOpen(openState),
          }}
        >
          <motion.nav
            animate={headerBar}
            className={`desktop-navigation-bar-menu-container ${className}`}
            css={css`
              & {
                position: fixed;
                top: 0;
                z-index: ${designSystem
                  .positioning
                  .behindBehindFirst};
                margin-top: 10px;
                height: 100px;
                padding: 15px;
              }
              &
                > .desktop-navigation-bar-hamburger-menu-container {
                height: 100%;
              }
            `}
          >
            <div
              className="desktop-navigation-bar-hamburger-menu-container"
              onClick={() => {
                setOpen(true);
                navBar.start({
                  x: 0,
                  transition: {
                    duration: 0.6,
                    ease: "easeOut",
                  },
                });
              }}
              role="presentation"
              onKeyDown={() => {
                setOpen(true);
                navBar.start({
                  x: 0,
                  transition: {
                    duration: 0.6,
                    ease: "easeOut",
                  },
                });
              }}
            >
              <HamburgerMenuIcon
                color={
                  backgroundIsWhite
                    ? undefined
                    : designSystem
                        .colors.brand
                        .primary
                }
                height="100%"
              />
            </div>
          </motion.nav>
          <nav
            css={css`
              & {
                display: ${open
                  ? "block"
                  : "none"};
                position: fixed;
                z-index: ${designSystem
                  .positioning
                  .behindFirst};
                top: 0;
                left: 0;
                width: 100vw;
                height: 100vh;
                transform-origin: left
                  center;
                background: rgba(
                  0,
                  0,
                  0,
                  0.35
                );
              }

              &
                > .desktop-navigation-bar-sidebar-container {
                position: relative;
                z-index: ${designSystem
                  .positioning.first};
                display: grid;
                justify-content: start;
                padding: 10px 0px;
                align-content: start;
                align-items: start;
                grid-template-rows:
                  repeat(3, auto)
                  1fr auto;
                float: left;
                height: 100%;
                background: ${designSystem
                  .colors.brand
                  .primary};
              }
              &
                .desktop-navigation-bar-logo-container {
                display: flex;
                justify-content: center;
                padding: 0px 20px;
              }
            `}
            className={`desktop-navigation-bar-sidebar-container-background ${
              className || ""
            }`}
            onClick={() => {
              navBar
                .start({
                  x: -200,
                  transition: {
                    duration: 0.8,
                    ease: "easeOut",
                  },
                })
                .then(() => {
                  setOpen(false);
                });
            }}
            onKeyDown={() => {
              navBar
                .start({
                  x: -200,
                  transition: {
                    duration: 0.8,
                    ease: "easeOut",
                  },
                })
                .then(() => {
                  setOpen(false);
                });
            }}
            role="presentation"
          >
            <motion.div
              className="desktop-navigation-bar-sidebar-container"
              animate={navBar}
              style={{
                x: -200,
              }}
            >
              <div className="desktop-navigation-bar-logo-container">
                <Logo />
              </div>
              <Spacer height="40px" />
              <FlexContainer
                direction="column"
                rowGap="10px"
              >
                {items.map(
                  ({
                    name,
                    href,
                    icon,
                    external,
                  }) => (
                    <NavigationBarItem
                      external={
                        external
                      }
                      key={name}
                      href={href}
                      icon={getIcon({
                        name: icon,
                      })}
                    >
                      {name}
                    </NavigationBarItem>
                  ),
                )}
              </FlexContainer>
              <div />
              <FlexContainer
                direction="column"
                rowGap="10px"
              >
                {footerItems.map(
                  ({
                    name,
                    href,
                    icon,
                  }) => (
                    <NavigationBarItem
                      key={name}
                      href={href}
                      icon={getIcon({
                        name: icon,
                      })}
                    >
                      {name}
                    </NavigationBarItem>
                  ),
                )}
              </FlexContainer>
            </motion.div>
          </nav>
        </DesktopNavigationBarContext.Provider>
      </>
    );
  },
);

NavigationBar.defaultProps = {
  backgroundIsWhite: true,
};

NavigationBar.displayName =
  "NavigationBar";
