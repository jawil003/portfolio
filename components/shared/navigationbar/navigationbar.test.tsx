import { render } from "@testing-library/react";
import { NavigationBar } from "./navigationbar";

describe("NavigationBar", () => {
  test("Create NavigationBar", () => {
    const testNavigationbar = render(
      <NavigationBar />,
    );

    expect(
      testNavigationbar.baseElement,
    ).toMatchSnapshot();
  });
});
