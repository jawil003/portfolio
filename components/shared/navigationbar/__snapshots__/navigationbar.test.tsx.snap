// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar Create NavigationBar 1`] = `
.emotion-0 {
  position: fixed;
  top: 0;
  z-index: 998;
  margin-top: 10px;
  height: 100px;
  padding: 15px;
}

.emotion-0>.desktop-navigation-bar-hamburger-menu-container {
  height: 100%;
}

.emotion-1 {
  display: none;
  position: fixed;
  z-index: 999;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  transform-origin: left center;
  background: rgba(
                    0,
                    0,
                    0,
                    0.35
                  );
}

.emotion-1>.desktop-navigation-bar-sidebar-container {
  position: relative;
  z-index: 1000;
  display: grid;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  padding: 10px 0px;
  -webkit-align-content: start;
  -ms-flex-line-pack: start;
  align-content: start;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  grid-template-rows: repeat(3, auto) 1fr auto;
  float: left;
  height: 100%;
  background: #f3f3f3;
}

.emotion-1 .desktop-navigation-bar-logo-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 0px 20px;
}

.emotion-2 {
  padding: 20px;
  cursor: pointer;
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  position: relative;
  overflow: visible;
}

.emotion-2>.logo-beta-label {
  background-color: #00739d;
  color: #f3f3f3;
  padding: 10px;
  font-size: 1.125rem;
  font-weight: 600;
  border-radius: 34px;
  position: absolute;
  bottom: 10px;
  right: -30px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex-wrap: no-wrap;
  -webkit-flex-wrap: no-wrap;
  -ms-flex-wrap: no-wrap;
  flex-wrap: no-wrap;
  margin: calc(10px / -2) calc(0px / -2);
}

.emotion-3>* {
  margin: calc(10px / 2) calc(0px / 2);
}

.emotion-4 {
  padding: 10px 20px;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #000;
  cursor: pointer;
}

.emotion-4>svg {
  fill: #000;
}

.emotion-4:hover {
  background-color: rgba(
                  0,
                  0,
                  0,
                  0.35
                );
}

.emotion-4>* {
  display: inline-block;
}

.emotion-5 {
  text-align: start;
  font-style: normal;
  margin: unset;
  -webkit-text-decoration: none;
  text-decoration: none;
  letter-spacing: normal;
  word-spacing: normal;
  color: #000;
  word-wrap: unset;
  padding: unset;
  word-break: normal;
}

.emotion-5 {
  font-weight: 600;
  font-size: 1em;
}

.emotion-7 {
  padding: 10px 20px;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #000;
  cursor: pointer;
}

.emotion-7>svg {
  fill: #000;
}

.emotion-7:hover {
  background-color: rgba(
                0,
                0,
                0,
                0.35
              );
}

.emotion-7>* {
  display: inline-block;
}

<body>
  <div>
    <nav
      class="desktop-navigation-bar-menu-container undefined emotion-0"
    >
      <div
        class="desktop-navigation-bar-hamburger-menu-container"
      >
        <svg
          class="hamburger-menu-icon undefined"
          clip-rule="evenodd"
          fill="#00739d"
          fill-rule="evenodd"
          height="100%"
          stroke-linejoin="round"
          stroke-miterlimit="2"
          viewBox="0 0 2481 2481"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
            fill="#00739d"
            transform="matrix(.7503 0 0 .7503 306.782 308.462) matrix(0 -1.1652 1.52426 0 -712.287 2973.76)"
          />
          <path
            d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
            fill="#00739d"
            transform="matrix(.7503 0 0 .7503 306.782 308.462) matrix(0 -1.1652 1.52426 0 -712.287 2076.49)"
          />
          <path
            d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
            fill="#00739d"
            transform="matrix(.7503 0 0 .7503 306.782 308.462) matrix(0 -1.1652 1.52426 0 -712.287 1179.26)"
          />
        </svg>
      </div>
    </nav>
    <nav
      class="desktop-navigation-bar-sidebar-container-background  emotion-1"
    >
      <div
        class="desktop-navigation-bar-sidebar-container"
        style="transform: translateX(-200px) translateZ(0);"
      >
        <div
          class="desktop-navigation-bar-logo-container"
        >
          <div
            class="logo-container emotion-2"
          >
            <svg
              fill-rule="evenodd"
              height="60px"
              viewBox="0 0 1925 1191"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
                fill="url(#a)"
                transform="matrix(.43892 -.2534 .3315 .57417 -346.149 44.635)"
              />
              <path
                d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
                fill="url(#b)"
                transform="matrix(.43892 -.2534 .3315 .57417 334.96 34.958)"
              />
              <path
                d="M966.527 638.508c0-105.526-112.075-191.2-250.12-191.2-138.044 0-250.119 85.674-250.119 191.2V1928.35c0 105.52 112.075 191.2 250.119 191.2 138.045 0 250.12-85.68 250.12-191.2V638.508z"
                fill="url(#c)"
                transform="matrix(.43892 .2534 .3315 -.57417 767.637 1145.681)"
              />
              <defs>
                <lineargradient
                  gradientTransform="matrix(8.55347 1718.32 -2247.83 6.53858 701.965 411.352)"
                  gradientUnits="userSpaceOnUse"
                  id="a"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    offset="0"
                    stop-color="#0078a3"
                  />
                  <stop
                    offset=".27"
                    stop-color="#00749e"
                  />
                  <stop
                    offset=".5"
                    stop-color="#00678d"
                  />
                  <stop
                    offset="1"
                    stop-color="#004a65"
                  />
                </lineargradient>
                <lineargradient
                  gradientTransform="matrix(8.55347 1718.32 -2247.83 6.53858 701.965 411.352)"
                  gradientUnits="userSpaceOnUse"
                  id="b"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    offset="0"
                    stop-color="#0078a3"
                  />
                  <stop
                    offset=".27"
                    stop-color="#00749e"
                  />
                  <stop
                    offset=".5"
                    stop-color="#00678d"
                  />
                  <stop
                    offset="1"
                    stop-color="#004a65"
                  />
                </lineargradient>
                <lineargradient
                  gradientTransform="matrix(8.55347 1718.32 -2247.83 6.53858 701.965 411.352)"
                  gradientUnits="userSpaceOnUse"
                  id="c"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    offset="0"
                    stop-color="#0078a3"
                  />
                  <stop
                    offset=".27"
                    stop-color="#00749e"
                  />
                  <stop
                    offset=".5"
                    stop-color="#00678d"
                  />
                  <stop
                    offset="1"
                    stop-color="#004a65"
                  />
                </lineargradient>
              </defs>
            </svg>
          </div>
        </div>
        <div
          style="height: 40px;"
        />
        <div
          class="emotion-3"
        >
          <a
            class="emotion-4"
          >
            <div
              style="display: inline-flex; align-items: center;"
            />
            <div
              style="width: 10px;"
            />
            <span
              class="emotion-5 emotion-6"
              color="#000"
              id="subtitle1"
            >
              Willkommen
            </span>
          </a>
          <a
            aria-hidden="true"
            class="emotion-7"
            href="https://willey3x37.hashnode.dev"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <div
              style="display: inline-flex; align-items: center;"
            />
            <div
              style="width: 10px;"
            />
            <span
              class="emotion-5 emotion-6"
              color="#000"
              id="subtitle1"
            >
              Veröffentlichungen
            </span>
          </a>
          <a
            aria-hidden="true"
            class="emotion-7"
            href="https://dribbble.com/willey3x37"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <div
              style="display: inline-flex; align-items: center;"
            />
            <div
              style="width: 10px;"
            />
            <span
              class="emotion-5 emotion-6"
              color="#000"
              id="subtitle1"
            >
              Designs
            </span>
          </a>
        </div>
        <div />
        <div
          class="emotion-3"
        />
      </div>
    </nav>
  </div>
</body>
`;
